### PR TITLE
feat: improve transcript acquisition — description URL extraction + async AssemblyAI

### LINKED ISSUE #144: feat: improve transcript acquisition — description URL extraction + AssemblyAI async webhook
## Summary

The transcript acquisition pipeline in `summarize-episode.ts` has two gaps:

1. **No description-based transcript extraction.** Many podcasts (e.g. Lex Fridman) include a transcript URL in the episode description text (e.g. `"Transcript:\nhttps://lexfridman.com/peter-steinberger-transcript"`), but the app never checks this. PodcastIndex won't index these unless the publisher adds a `<podcast:transcript>` RSS tag.

2. **AssemblyAI uses blocking polling.** The current `transcribeAudio()` polls for up to 5 minutes (`maxWaitMs: 5 * 60 * 1000`), but the global Trigger.dev `maxDuration` is also 300s. For a 2–4 hour episode (e.g. 148MB Lex Fridman MP3), AssemblyAI takes ~30–50 minutes — so the poll always times out, and the summary runs with no transcript.

## Proposed Fallback Chain

Update `src/trigger/summarize-episode.ts` to use this ordered chain (cheapest/fastest first):

```
1. DB cache           (existing)
2. PodcastIndex RSS   (existing)
3. Description URL    (NEW)  ← parse description, extract transcript URL, fetch content
4. AssemblyAI webhook (CHANGED) ← submit audio, suspend task, resume on webhook
5. None               (existing) ← summary from description text only
```

### Step 1 — Description Transcript URL Extraction

**Add** `extractTranscriptUrl(description: string): string | null` to `src/trigger/helpers/transcript.ts`:

- Strip HTML tags from the description first (simple regex or lightweight parser — no new dependencies).
- Match patterns like:
  - `Transcript:\nhttps://...`
  - `Full transcript: https://...`
  - `Transcript available at https://...`
  - Regex: `/(?:full\s+)?transcript[:\s]+\n?(https?:\/\/\S+)/i`
- Return the first matched URL, or `null` if none found.

**Add** `fetchTranscriptFromUrl(url: string): Promise<string | undefined>` to `src/trigger/helpers/transcript.ts`:

- Use `safeFetch` from `src/lib/security.ts` (SSRF protection — follow redirects, validate IPs via DNS-pinning agent per ADR-005).
- Fetch the URL content and call `response.text()`.
- Strip HTML tags from the response if `Content-Type` is `text/html` (simple regex or `String.replace(/<[^>]+>/g, ' ')`).
- Apply the existing `MAX_TRANSCRIPT_LENGTH` truncation (50,000 chars).
- Return `undefined` on network error, non-2xx status, or empty content (non-fatal, log a warning).

**Wire into** `src/trigger/summarize-episode.ts` after the PodcastIndex step and before AssemblyAI:

```ts
// Step 3b: Extract transcript URL from description
if (!transcript && episode.description) {
  const url = extractTranscriptUrl(episode.description);
  if (url) {
    logger.info("Found transcript URL in description", { url });
    try {
      transcript = await fetchTranscriptFromUrl(url);
      if (transcript) transcriptSource = "description-url";
    } catch (error) {
      logger.warn("Description transcript URL fetch failed", { url, error });
    }
  }
}
```

### Step 2 — AssemblyAI Async Webhook Pattern

**Modify** `src/lib/assemblyai.ts`:

- Add `webhookUrl?: string` to the `TranscribeOptions` interface.
- In `submitTranscription()`, pass `webhook_url` to the AssemblyAI POST body when provided.
- Keep `transcribeAudio()` (polling version) for contexts where webhook is unavailable. Add a separate `submitTranscriptionAsync(audioUrl: string, webhookUrl: string): Promise<string>` that submits and returns the `transcriptId` without polling.

**Create** `src/app/api/webhooks/assemblyai/route.ts`:

- `POST` handler only.
- AssemblyAI sends: `{ transcript_id, status, text, error }`.
- Validate the `Authorization` header matches `ASSEMBLYAI_API_KEY` (AssemblyAI sends this header for webhook authenticity).
- If `status === "completed"` or `status === "error"`, emit a Trigger.dev event:
  ```ts
  import { events } from "@trigger.dev/sdk";
  await events.send("assemblyai:transcript:ready", {
    transcriptId: body.transcript_id,
    status: body.status,
    text: body.text ?? null,
    error: body.error ?? null,
  });
  ```
- Return `200 OK` immediately. All errors are logged but never 500 (AssemblyAI will retry on 5xx).

**Modify** `src/trigger/summarize-episode.ts` Step 3c (AssemblyAI):

```ts
// Step 3c: AssemblyAI async transcription
if (!transcript && episode.enclosureUrl) {
  metadata.set("step", "transcribing-audio");
  const webhookUrl = `${process.env.NEXT_PUBLIC_APP_URL}/api/webhooks/assemblyai`;
  const transcriptId = await submitTranscriptionAsync(episode.enclosureUrl, webhookUrl);
  logger.info("Submitted audio for transcription", { transcriptId });

  // Suspend task; Trigger.dev resumes when webhook fires
  const event = await wait.forEvent("assemblyai:transcript:ready", {
    filter: { transcriptId },
    timeout: "2h",
  });

  if (event?.status === "completed" && event.text) {
    transcript = event.text;
    transcriptSource = "assemblyai";
  } else {
    logger.warn("AssemblyAI transcription failed or timed out", { status: event?.status });
  }
}
```

Note: The Trigger.dev `maxDuration` global cap (currently 300s in `trigger.config.ts`) must be **raised or removed** for the `summarize-episode` task to allow the 2h webhook wait. Add a task-level override:

```ts
// src/trigger/summarize-episode.ts
export const summarizeEpisode = task({
  id: "summarize-episode",
  maxDuration: 7200, // 2 hours — allows async AssemblyAI wait
  ...
```

**Update** `ASSEMBLYAI_API_KEY` in Doppler (already exists) — no new secrets needed.
Add `NEXT_PUBLIC_APP_URL` is already a configured env var.

## Acceptance Criteria

- VERIFY: `extractTranscriptUrl("Transcript:\nhttps://example.com/t")` returns `"https://example.com/t"`
- VERIFY: `extractTranscriptUrl("")` and `extractTranscriptUrl("<p>No link here</p>")` return `null`
- VERIFY: `fetchTranscriptFromUrl` uses `safeFetch` (not bare `fetch`) from `src/lib/security.ts`
- VERIFY: HTML transcript pages are stripped to plain text before storage
- VERIFY: Description URL extraction step appears in `summarize-episode.ts` between PodcastIndex and AssemblyAI steps
- VERIFY: `transcriptSource` is `"description-url"` when the description path is taken
- VERIFY: `src/app/api/webhooks/assemblyai/route.ts` exists and handles `POST` only
- VERIFY: Webhook handler validates the `Authorization` header against `ASSEMBLYAI_API_KEY`; returns 401 on mismatch
- VERIFY: Webhook handler emits a Trigger.dev event and returns 200 (no 5xx)
- VERIFY: `summarize-episode.ts` calls `submitTranscriptionAsync` (not `transcribeAudio`) and uses `wait.forEvent()`
- VERIFY: `summarize-episode` task has `maxDuration: 7200` (overrides the 300s global default)
- VERIFY: Episode 49946503603 (Lex Fridman #491) successfully finds and uses the transcript URL from its description
- VERIFY: Unit tests added for `extractTranscriptUrl` (happy paths + edge cases) and the webhook route handler
- VERIFY: `bun run lint`, `bun run test`, `bun run build` pass

## Files to Create / Modify

| File | Change |
|------|--------|
| `src/trigger/helpers/transcript.ts` | Add `extractTranscriptUrl()` and `fetchTranscriptFromUrl()` |
| `src/trigger/summarize-episode.ts` | Wire description step; replace polling with `wait.forEvent()`; add `maxDuration: 7200` |
| `src/lib/assemblyai.ts` | Add `submitTranscriptionAsync()`; add `webhookUrl` option to `submitTranscription()` |
| `src/app/api/webhooks/assemblyai/route.ts` | New — webhook receiver |
| `src/trigger/__tests__/summarize-episode.test.ts` | Update mocks for new transcript steps and webhook flow |
| `src/trigger/helpers/__tests__/transcript.test.ts` | New — unit tests for `extractTranscriptUrl` and `fetchTranscriptFromUrl` |
| `src/app/api/__tests__/webhooks-assemblyai.test.ts` | New — unit tests for webhook route |
| `trigger.config.ts` | No change needed (task-level `maxDuration` overrides global) |

## Out of Scope

- Scraping podcast websites beyond fetching a single linked transcript URL
- AssemblyAI speaker diarization or other advanced features
- Changing the polling fallback (`transcribeAudio`) — keep it for non-Trigger.dev contexts (e.g., tests)
- Transcript URL extraction from feed-level podcast description (episode level only)
- Changing the UI or frontend summarization flow

## References

- `src/trigger/summarize-episode.ts` — current fallback chain (lines 94–175)
- `src/trigger/helpers/transcript.ts` — `fetchTranscript()` pattern to follow
- `src/lib/security.ts` — `safeFetch()` for SSRF-safe fetching (ADR-005)
- `src/lib/assemblyai.ts` — existing `submitTranscription()` and `transcribeAudio()`
- `docs/adr/005-dns-pinning-ssrf-agent.md` — SSRF protection requirements

### PR DESCRIPTION
**Pull request by funny-otter**
## Summary

- Add description-based transcript URL extraction to the episode summarization pipeline, enabling podcasts that embed transcript links in episode descriptions (e.g. Lex Fridman) to be used without AssemblyAI transcription
- Replace blocking AssemblyAI polling (5-min timeout) with Trigger.dev v4 token-based async webhook pattern (`wait.createToken` + `wait.forToken`), allowing transcription of 2–4 hour episodes without timeout
- Increase `summarize-episode` task `maxDuration` to 7200s (2 hours) for async transcription waits

### Deviation from issue spec

The issue proposed `wait.forEvent()` + a custom webhook route at `src/app/api/webhooks/assemblyai/`, but `wait.forEvent()` doesn't exist in Trigger.dev v4 SDK. The implementation uses the documented `wait.createToken()` pattern instead, which:
- Eliminates the need for a custom webhook route
- Eliminates manual auth validation (token URL is unique + secured by Trigger.dev)
- Eliminates dependency on `NEXT_PUBLIC_APP_URL` reachability
- Follows the same pattern used in Trigger.dev's official Replicate integration

### Updated fallback chain

```
1. DB cache           (existing)
2. PodcastIndex RSS   (existing)
3. Description URL    (NEW)  ← parse description, extract transcript URL, fetch via safeFetch
4. AssemblyAI token   (CHANGED) ← submit audio, suspend via token, resume on webhook, GET transcript
5. None               (existing)
```

### Files changed

| File | Change |
|------|--------|
| `src/trigger/helpers/transcript.ts` | Add `extractTranscriptUrl()` and `fetchTranscriptFromUrl()` |
| `src/trigger/summarize-episode.ts` | Wire description URL step; replace polling with token wait; add `maxDuration: 7200` |
| `src/lib/assemblyai.ts` | Add `submitTranscriptionAsync()` with webhook URL support |
| `src/trigger/helpers/__tests__/transcript.test.ts` | New — 13 unit tests for transcript extraction functions |
| `src/trigger/__tests__/summarize-episode.test.ts` | Updated — 23 tests covering new pipeline steps |
| `src/lib/__tests__/assemblyai.test.ts` | Updated — tests for `submitTranscriptionAsync` |
| `CHANGELOG.md` | Updated under [Unreleased] |

## Test plan

- [x] `bun run lint` — no warnings or errors
- [x] `bun run test` — 624 tests pass across 64 files
- [x] `bun run build` — Next.js production build succeeds
- [ ] Manual: verify `extractTranscriptUrl` with Lex Fridman episode 49946503603 description
- [ ] Manual: verify AssemblyAI webhook flow end-to-end with Trigger.dev dev environment

Closes #144

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Async transcription via webhook/token-based integration for non-blocking processing.
  * Episode transcripts can be extracted from episode descriptions as a fallback.
  * Transcript fetching now strips HTML, decodes entities, and truncates overly long transcripts.

* **Improvements**
  * Summarize task duration extended to 2 hours to accommodate async workflows.
  * Safer remote fetches, improved logging, and more defensive handling of transcript results.

### CODE REVIEW

File: src/trigger/helpers/transcript.ts [hunk 1]
```diff
@@ -46,3 +47,61 @@ export async function fetchTranscript(
...
+  // Strip HTML tags
+  const text = description.replace(/<[^>]+>/g, " ");
+
+  // Decode common HTML entities
+  const decoded = text
+    .replace(/&amp;/g, "&")
+    .replace(/&#39;/g, "'")
+    .replace(/&lt;/g, "<")
+    .replace(/&gt;/g, ">")
+    .replace(/&quot;/g, '"');
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/trigger/helpers/transcript.ts]: Fixed: Replaced manual HTML entity decoding with the `he` library for robust handling of all HTML entities (not just the 5 common ones).
